packages/runner/lib/usage.ts

```diff
@@ -0,0 +1,57 @@
+import os from 'os';
+import { createActivityLogMessage } from '@nangohq/shared';
+
+interface Tracked {
+    activityLogId: number;
+    environmentId: number;
+}
+
+const MEMORY_WARNING_PERCENTAGE_THRESHOLD = 70;
+let lastMemoryReportDate: Date | null = null;
+const tracked: Map<number, Tracked> = new Map<number, Tracked>();
+
+export function track(activityLogId: number, environmentId: number) {
+    tracked.set(activityLogId, { activityLogId, environmentId });
+}
+
+export function untrack(activityLogId: number) {
+    tracked.delete(activityLogId);
+}
+
+export function monitorUsage(): void {
+    checkMemoryUsage();
+}
+
+function checkMemoryUsage(): void {
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    setInterval(async () => {
+        const rss = process.memoryUsage().rss;
+        const total = os.totalmem();
+        const memoryUsagePercentage = (rss / total) * 100;
+
+        if (memoryUsagePercentage > MEMORY_WARNING_PERCENTAGE_THRESHOLD) {
+            await report(memoryUsagePercentage);
+        }
+    }, 1000);
```

Comment:
A floating setInterval will block a graceful shutdown, it needs to be stored and closed properly